### ISSUE TITLE
[Feature] Allow customized, fixed viewer bound in interactive mode

### ISSUE THREAD

**Issue by Jianye-Xu** [proroklab/VectorizedMultiAgentSimulator#92]
Current, VMAS seemly does not support a customized, fixed viewer bound in interactive mode. This feature would be useful if the origin $(0,0)$ of a user's map is not centered. In my case, the origin of my map is located at the lower left corner, meaning that the $x$-range of my map is `[0, world_x_dim]`, and the $y$-range of my map is `[0, world_y_dim]`, instead of `[-world_x_dim/2, world_x_dim/2]` and `[-world_y_dim/2, world_y_dim/2]`. In addition, I do not want to shift my map since I am using an HD map with dense data.

My current workaround is to add an attribute `viewer_bound` in my **Scenario** class's `make_world` function:
```
self.viewer_bound = torch.tensor(
    [0, world_x_dim, 0, world_y_dim],
    device=device,
    dtype=torch.float32,
)
```
and then modify the source code of file `/vmas/simulator/environment/environment.py` around [this line](https://github.com/proroklab/VectorizedMultiAgentSimulator/blob/146d0c1b36493d2a1eda2506ea60d17d70486a56/vmas/simulator/environment/environment.py#L700) from
```
self.viewer.set_bounds(
    -cam_range[X],
    cam_range[X],
    -cam_range[Y],
    cam_range[Y],
)
```
to
```
if hasattr(self.scenario, "viewer_bound"):
    self.viewer.set_bounds(
        self.scenario.viewer_bound[0], # left boundary
        self.scenario.viewer_bound[1], # right boundary
        self.scenario.viewer_bound[2], # bottom boundary
        self.scenario.viewer_bound[3], # top boundary
    )
else:
    self.viewer.set_bounds(
        -cam_range[X],
        cam_range[X],
        -cam_range[Y],
        cam_range[Y],
    )
```
This way, if `self.scenario` contains the user-given attibute `viewer_bound`, the viewer bound will be set accordingly; otherwise, the funtionality is untouched.

I would appreciate it if anyone could point out if there is an easier way to achieve this. If not, I think it would be nice if the developer would add such a feature to VMAS.

**Comment by matteobettini** [comment]
Hey!

Thanks so much for the comment. I think I get your request.

In other words you want to have the center of the render image in a custom point and not in [0,0].

My suggestion is this:

- have an attiribute in the scenario called `render_center` which defaults to `[0,0]` and is used only when `agent_index_focus=None` in `render()`
- this param would be added to the line you linked like so 
```python
self.viewer.set_bounds(
        -cam_range[X] + render_center[X],
        cam_range[X]  + render_center[X],
        -cam_range[Y] + render_center[Y],
        cam_range[Y] + render_center[Y],
    )
```
- this way you can override this and set it to `[ world_x_dim/2, world_y_dim/2]`
- in addition to your solution this would also zoom out in case any agent goes off the map (which you can easily prevent by setting the world bounds)

What fo you think?

**Comment by matteobettini** [comment]
I am down to enable this feature.

With this added you would have control over:
- scenario.render_center
- scenario.viewer_size
- scenario.zoom

Which should let you do any rendering you want I think

**Comment by matteobettini** [comment]
Another very naive solution is just to shift the map to the origin of the render

**Comment by Jianye-Xu** [comment]
Thanks for your suggestion! 

I tried it and think it would be a good solution. There are two small issues I encountered:
- There are blank margins around my map. I tried to adjust `scenario.viewer_size` but it did not change anything.
- The viewer bound is dynamically changing, which is  unexpected since I am using a closed map and agents will be reset once they across the map boundaries.

Do you have any ideas to fix these two issues?

This is a video demo, which hopefully could help in understanding:

https://github.com/proroklab/VectorizedMultiAgentSimulator/assets/83649047/45ecb35b-a10d-46b8-ae76-3721f691751e

**Comment by matteobettini** [comment]
Wow that scenario looks so cool!

- for the first issue: viewer_size only determines the aspect ratio, if your map is a square i would not change that, have you tried changing the zoom?
- For the second issue: the bounds should dynamically change only if an agent gets within $\epsilon$ of getting out of sight. I thought this should not happen for a reasonable zoom, but we can have anotehr param scenario.fixed_zoom to make it fixed

**Comment by Jianye-Xu** [comment]
Thanks!

- Yes, I also tried to change `scenario.viewer_zoom` but it did not change anything.
- Yeah this is also want I expected. But somehow the bounds still changed dynamically when some agents moved, not to mention that the map was already zoomed out a lot.

My setups are as follows:
```
world_x_dim = 4.5  # [m]
world_y_dim = 4.0 # [m]

self.viewer_zoom = 1.0 # The adjustments of this parameter did not influence anything in my case
self.viewer_size = (world_x_dim * 200, world_y_dim * 200)
```

Do you have any clues?

**Comment by matteobettini** [comment]
Could you try out freshly in #93 please?

**Comment by Jianye-Xu** [comment]
Thank you! I tried it and there are no unwanted blank margins now. Most of the time, it works very well. However, when all agents are close to each other, the view zooms in significantly (see the Fig. below), which leads to the loss of the map's overview. This is beneficial as it highlights the agents rather than the map's overview. However, sometimes I think it would be nice to have a fixed viewer bound, for example, when I put the evaluation videos of different policies together to intuitively compare their performance, the dynamically changing zoom may be somehow distracting.

What do you think?

<img width="909" alt="image" src="https://github.com/proroklab/VectorizedMultiAgentSimulator/assets/83649047/e35dad84-f0be-4f34-a4a1-29f7c5db88a1">

**Comment by matteobettini** [comment]
The dynamical zoom should not happen if you set scenario.viewer_zoom to a high enough value.

try incrementing that parameter up to a certain value and you should get what you are looking for

Otherwise let me know and i will investigate further

**Comment by Jianye-Xu** [comment]
Thanks for pointing this out. This is what exactly what I am looking for! 

I think you can close the issue now.

**Comment by matteobettini** [comment]
Very glad it worked!

From the rendering this scenario looks super exciting! Would love to add it to vmas if you feel like contributing it one day!

**Comment by Jianye-Xu** [comment]
Sure, I would definitely love to! I will let you know when the time comes!
